![LogoUC3M](https://upload.wikimedia.org/wikipedia/commons/4/47/Acronimo_y_nombre_uc3m.png)

Aprendizaje Automático · Grado en Ingeniería Informática · Curso 2022/23
---
**Cuaderno** - Práctica 2: Predicción del abandono (burnout) de empleados
<br><br>
***Grupo Reducido 82 - Grupo de laboratorio 13***

# **Carga de datos**
Para comenzar, importaremos las bibliotecas que necesitamos:

+ `biblioteca`: Definición.

 ################################### <br>
**(AQUÍ VAN LAS BIBLIOTECAS NECESARIAS)<br>
(AQUÍ VAN LAS BIBLIOTECAS NECESARIAS)<br>
(AQUÍ VAN LAS BIBLIOTECAS NECESARIAS)<br>
(AQUÍ VAN LAS BIBLIOTECAS NECESARIAS)<br>
(AQUÍ VAN LAS BIBLIOTECAS NECESARIAS)**<br>
 ################################### <br>

Una vez instaladas las bibliotecas necesarias, importamos **Numpy** y **Pandas** ya que los necesitaremos para el desarrollo de la práctica.
Destacar que nos referiremos a **Numpy** como *np* y a **Pandas** como *pd* por simplicidad.

In [8]:
import numpy as np
import pandas as pd

A continuación, cargaremos los datos necesarios. Dado que somos el **grupo 13**, usaremos los dos conjuntos de datos correspondientes.</br></br>Los datos serán divididos en un conjunto de **train** para entrenar y ajustar hiperparámetros, y un conjunto de **test** en el que evaluaremos las distintas posibilidades que se probarán en la práctica.</br></br>

In [9]:
import pickle

# Datos disponibles.
with open('attrition_available_13.pkl', 'rb') as f:
    data = pickle.load(f)

df = pd.DataFrame(data)

print(df)

# Mostramos la información del conjunto de datos.
print(f"El conjunto de datos tiene {len(df)} instancias.")

           hrs  absences  JobInvolvement  PerformanceRating  \
1     7.315971      13.0             2.0                4.0   
6     6.450877      17.0             3.0                4.0   
13         NaN      14.0             NaN                3.0   
28         NaN       NaN             2.0                3.0   
30    9.662808      11.0             NaN                3.0   
...        ...       ...             ...                ...   
4405  8.316921       6.0             3.0                3.0   
4406  5.897197       8.0             2.0                3.0   
4407  7.149526      18.0             3.0                NaN   
4408  9.187612       NaN             NaN                3.0   
4409  6.511790      17.0             NaN                3.0   

      EnvironmentSatisfaction  JobSatisfaction  WorkLifeBalance   Age  \
1                         3.0              2.0              NaN   NaN   
6                         1.0              3.0              1.0  28.0   
13                      

A continuación, separamos la **matriz de atributos** (**X**) y el **vector de la variable de respuesta** *salida* (**y**).

In [11]:
# Datos.
X = df.drop('Attrition', axis=1)

# Etiquetas.
y = df.Attrition

print(y)

1       Yes
6       Yes
13      Yes
28      Yes
30      Yes
       ... 
4405     No
4406     No
4407     No
4408     No
4409     No
Name: Attrition, Length: 4410, dtype: object


# **Análisis Exploratorio de Datos (EDA)**
Antes de comenzar, destacar que dado que los resultados deben ser reproducibles, hemos fijado la **semilla de números aleatorios** en los lugares adecuados. Para ello, hemos seleccionado como semilla el número del grupo de prácticas (*13*).</br></br>
Para poder realizar el análisis exploratorio de datos, llevaremos a cabo los siguientes pasos:

+ Número de instancias y atributos.
+ Tipo de atributos (numéricos o categóricos).
+ Columnas constantes o innecesarias.
+ Proporción de *missing values*.
+ Tipo de problema (clasificación o regresión).
+ Desbalance.

## **Número de instancias y atributos**

## **Número de instancias y atributos**
Los **datos** cuentan con 30 atributos y la variable de respuesta *Attrition*. En total, contamos con un total de **4410 instancias**.</br></br>

A continuación, profundizaremos más en los atributos, su significado y nomenclatura.</br></br>

Como podemos ver en la siguiente tabla, contamos con **30 atributos** diferentes, cada uno de ellos con su respectivo nombre (identificador) y con su unidad concreta. Estos atributos corresponden con **variables sobre empleados de la empresa recopilados por el departamento de recursos humanos**. El último valor que podemos observar es la predicción de abandono de empleados (burnout) o, de otra forma, la variable de salida de nuestros datos.</br></br>

| Columna | Descripción | Tipo de dato |
| :- | :- | -: |
| `hrs` | Número de horas trabajadas. | $float64$ |
| `absences` | Descripción. | $float64$ |
| `JobInvolvement` | Descripción. | $float64$ |
| `PerformanceRating` | Descripción. | $float64$ |
| `EnvironmentSatisfaction`	 | Descripción. | $float64$ |
| `JobSatisfaction`	 | Descripción. | $float64$ |
| `WorkLifeBalance`	 | Descripción. | $float64$ |
| `Age` | Descripción. | $float64$ |
| `BusinessTravel` | Descripción. | $object$ |
| `Department` | Descripción.	| $object$ |
| `DistanceFromHome` | Descripción. | $float64$ |
| `Education` | Descripción. | $int64$ |
| `EducationField` | Descripción. | $object$ |
| `EmployeeCount` | Descripción. | $float64$ |
| `EmployeeID` | Descripción. | $int64$ |
| `Gender` | Descripción. | $object$ |
| `JobLevel` | Descripción. | $float64$ |
| `JobRole` | Descripción. | $object$ |
| `MaritalStatus` | Descripción. | $object$ |
| `MonthlyIncome` | Descripción. | $float64$ |
| `NumCompaniesWorked` | Descripción. | $float64$ |
| `Over18` | Descripción. | $object$ |
| `PercentSalaryHike` | Descripción. | $float64$ |
| `StandardHours` | Descripción. | $float64$ |
| `StackOptionLevel` | Descripción. | $float64$ |
| `TotalWorkingYears` | Descripción. | $float64$ |
| `TrainingTimesLastYear` | Descripción. | $float64$ |
| `YearsAtCompany` | Descripción. | $float64$ |
| `YearsSinceLastPromotion` | Descripción. | $float64$ |
| `YearsWithCurrManager` | Descripción. | $float64$ |
| `Attrition` | Descripción. | $object$ |


## **Tipo de atributos**

Dentro de los diferentes tipos de atributos de los datos podemos encontrar los categóricos, numéricos u ordinales.</br></br>

En el caso de nuestros datos, podemos ver como claramente como cada uno de ellos es de un tipo diferente como indicamos en la siguiente tabla.</br></br>

| Columna | Tipo de atributo |
| :- | -: |
| `hrs` | $(tipo)$ |
| `absences` | $(tipo)$ |
| `JobInvolvement` | $(tipo)$ |
| `PerformanceRating` | $(tipo)$ |
| `EnvironmentSatisfaction` | $(tipo)$ |
| `JobSatisfaction` | $(tipo)$ |
| `WorkLifeBalance` | $(tipo)$ |
| `Age` | $(tipo)$ |
| `BusinessTravel` | $(tipo)$ |
| `Department` | $(tipo)$ |
| `DistanceFromHome` | $(tipo)$ |
| `Education` | $(tipo)$ |
| `EducationField` | $(tipo)$ |
| `EmployeeCount` | $(tipo)$ |
| `EmployeeID` | $(tipo)$ |
| `Gender` | $(tipo)$ |
| `JobLevel` | $(tipo)$ |
| `JobRole` | $(tipo)$ |
| `MaritalStatus` | $(tipo)$ |
| `MonthlyIncome` | $(tipo)$ |
| `NumCompaniesWorked` | $(tipo)$ |
| `Over18` | $(tipo)$ |
| `PercentSalaryHike` | $(tipo)$ |
| `StandardHours` | $(tipo)$ |
| `StackOptionLevel` | $(tipo)$ |
| `TotalWorkingYears` | $(tipo)$ |
| `TrainingTimesLastYear` | $(tipo)$ |
| `YearsAtCompany` | $(tipo)$ |
| `YearsSinceLastPromotion` | $(tipo)$ |
| `YearsWithCurrManager` | $(tipo)$ |
| `Attrition` | $(tipo)$ |

<br><br>
Por tanto, dado que no todos los datos con los que vamos a tratar son atributos numéricos, tendremos que transformar las variables categóricas en **one-hot-encoding** y las variables ordinales en numéricas.

## **Columnas constantes o innecesarias**

## **Proporción de missing values**

A continuación, comprobaremos los **missings values** que existen en nuestros datos. Para ello, tomaremos el dataframe de datos creado anteriormente y haremos una suma por atributo de los missing values y lo compararemos con el número total de valores. Tras esto, crearemos una tabla para poder ver de manera clara las proporciones de missing values de cada atributo.

In [20]:
# Función que cuenta el número de missing values por atributos,
def missing_values_table(df):
    mis_val = df.isnull().sum()
    mis_val_percent = 100 * df.isnull().sum() / len(df)
    mis_val_percent_rounded = mis_val_percent.round(2).astype(str) + '%'
    mis_val_table = pd.concat([mis_val, mis_val_percent_rounded], axis=1)
    mis_val_table_ren_columns = mis_val_table.rename(
    columns = {0 : 'Missing Values', 1 : 'Proporción'})
    return mis_val_table_ren_columns
  
# Generamos las proporciones de missing values.
missing_values_table(df)

Missing Values Proporción
hrs                                 943     21.38%
absences                            755     17.12%
JobInvolvement                      884     20.05%
PerformanceRating                   962     21.81%
EnvironmentSatisfaction             811     18.39%
JobSatisfaction                     904      20.5%
WorkLifeBalance                     935      21.2%
Age                                 825     18.71%
Attrition                             0       0.0%
BusinessTravel                      852     19.32%
Department                            0       0.0%
DistanceFromHome                    884     20.05%
Education                           910     20.63%
EducationField                      769     17.44%
EmployeeCount                       897     20.34%
EmployeeID                          964     21.86%
Gender                                0       0.0%
JobLevel                            779     17.66%
JobRole                               0       0.0%
MaritalStatus                       932     21.13%
MonthlyIncome                       755     17.12%
NumCompaniesWorked                  816      18.5%
Over18                              822     18.64%
PercentSalaryHike                   868     19.68%
StandardHours                       708     16.05%
StockOptionLevel                    755     17.12%
TotalWorkingYears                   915     20.75%
TrainingTimesLastYear               790     17.91%
YearsAtCompany                      729     16.53%
YearsSinceLastPromotion               0       0.0%
YearsWithCurrManager                747     16.94%

## **Tipo de problema**

## **Desbalance**

## **Análisis de atributos**

# **Construcción de modelos**

 ################################### <br>
**(ENUNCIADO - BORRAR LUEGO)<br>
(ENUNCIADO - BORRAR LUEGO)<br>
(ENUNCIADO - BORRAR LUEGO)<br>
(ENUNCIADO - BORRAR LUEGO)<br>
(ENUNCIADO - BORRAR LUEGO)**<br>
 ################################### <br>

**Construcción de modelos**: para esta práctica usaremos **LogisticRegression** como método base (sin ajustar hiper-parámetros) y Boosting como método avanzado (ajustando hiper-parámetros), a elegir. Es importante realizar los preprocesos que los datos necesiten, usando preferentemente **pipelines**. Como método de boosting, se puede elegir uno de entre los métodos de **boosting** disponibles en scikit-learn.
<br><br>
**EXTRA**: Si además se usa uno de entre las **librerías externas** xgboost, lightgbm o catboost, se pueden sacar +0.35 puntos adicionales.

# **Mejora de resultados**

 ################################### <br>
**(ENUNCIADO - BORRAR LUEGO)<br>
(ENUNCIADO - BORRAR LUEGO)<br>
(ENUNCIADO - BORRAR LUEGO)<br>
(ENUNCIADO - BORRAR LUEGO)<br>
(ENUNCIADO - BORRAR LUEGO)**<br>
 ################################### <br>

Usando algún **método de selección de atributos de tipo filter (SelectKBest)** de entre los disponibles en sklearn **(f_classif, mutual_info_classif o chi2)**, comprobad si se pueden mejorar los resultados del apartado anterior y extraer conclusiones sobre qué atributos son más importantes, al menos de acuerdo a estos métodos.